lib/sources/bulk_assembly.rb

```diff
@@ -0,0 +1,130 @@
+
+        raise
+      end
+    end
+
+    def create_sources(sources)
+      sources&.map do |source|
+        # get the source type by ID or by type string
+        srct = SourceType.find_by(:id => source.delete(:source_type_id)) || SourceType.find_by(:name => source.delete(:type))
+
+        raise ActiveRecord::ActiveRecordError, "Source Type not found" if srct.nil?
+
+        Source.create!(source.merge!(:source_type => srct))
+      end
+    end
+
+    def create_endpoints(endpoints, resources)
+      endpoints&.map do |endpoint|
+        src = find_resource(resources, :sources, endpoint.delete(:source_name))
+
+        Endpoint.create!(endpoint.merge!(:source_id => src.id))
+      end
+    end
+
+    def create_applications(applications, resources)
+      applications&.map do |app|
+        src = find_resource(resources, :sources, app.delete(:source_name))
+        # Get the application by id or lookup by type string
+        appt = ApplicationType.find_by(:id => app.delete(:application_type_id)) || get_application_type(app.delete(:type))
+
+        ::Application.create!(app.merge!(:source_id => src.id, :application_type_id => appt.id))
+      end
+    end
+
+    def create_authentications(authentications, resources)
+      authentications&.map do |auth|
+        ptype = auth.delete(:resource_type)
+        pname = auth.delete(:resource_name)
+
+        # complicated logic here - since the source/endpoint/application can all
+        # be looked up differently
+        parent = case ptype
+                 when "source"
+                   find_resource(resources, :sources, pname)
+                 when "endpoint"
+                   find_resource(resources, :endpoints, pname, :host)
+                 when "application"
+                   # we have to look up the application by id before jumping
+                   # into looking for the current resources since it matches by
+                   # type
+                   ::Application.find_by(:id => pname) || find_resource(resources, :applications, get_application_type(pname), :application_type)
+                 end
+
+        Authentication.create!(auth.merge!(:resource => parent)).tap do |newauth|
+          # create the application_authentication relation if the parent was an
+          # application.
+          ApplicationAuthentication.create!(:authentication => newauth, :application => parent) if ptype == "application"
+        end
+      end
+    end
+
+    private
+
+    def find_resource(resources, rtype, rname, field = :name)
+      # use the safe operator in the case of creating a subresource on an
+      # existing source
+      parent = resources[rtype]&.detect { |resource| resource.send(field) == rname }
+
+      # if the parent is a source, it's possible that it was already created in
+      # the db so we need to try and look it up
+      if parent.nil?
+        case rtype
+        when :sources
```

Comment:
same as above, should match the resource_type values format, for better understandability